pkg/registry/service/rest.go

```diff
@@ -45,21 +46,104 @@ type REST struct {
 	machines    minion.Registry
 	endpoints   endpoint.Registry
 	portals     ipallocator.Interface
+	portMgr     *portAllocator
 	clusterName string
 }
 
 // NewStorage returns a new REST.
 func NewStorage(registry Registry, machines minion.Registry, endpoints endpoint.Registry, portals ipallocator.Interface,
-	clusterName string) *REST {
+	publicServicePorts *portAllocator, clusterName string) *REST {
+
 	return &REST{
 		registry:    registry,
 		machines:    machines,
 		endpoints:   endpoints,
 		portals:     portals,
+		portMgr:     publicServicePorts,
 		clusterName: clusterName,
 	}
 }
 
+// Encapsulates the semantics of a port allocation 'transaction'
+// it is better to leak ports than to double-allocate them
+// so we allocate immediately, but defer release
+// on commit we best-effort release the deferred releases
+// on rollback we best-effort release any allocations we did
+type portAllocationOperation struct {
+	pa              *portAllocator
+	allocated       []int
+	releaseDeferred []int
+	ShouldRollback  bool
+}
+
+func (op *portAllocationOperation) Init(pa *portAllocator) {
```

Comment:
> On May 15, 2015, at 11:46 AM, Tim Hockin <notifications@github.com> wrote:
> 
> In pkg/registry/service/rest.go:
> 
> >  		clusterName: clusterName,
> >  	}
> >  }
> >  
> > +// Encapsulates the semantics of a port allocation 'transaction'
> > +// it is better to leak ports than to double-allocate them
> > +// so we allocate immediately, but defer release
> > +// on commit we best-effort release the deferred releases
> > +// on rollback we best-effort release any allocations we did
> > +type portAllocationOperation struct {
> > +	pa              *portAllocator
> > +	allocated       []int
> > +	releaseDeferred []int
> > +	ShouldRollback  bool
> > +}
> > +
> > +func (op *portAllocationOperation) Init(pa *portAllocator) {
> The assumption was that since the port is part of spec and we're not doing a Binding yet, the user CAN request a particular port. It's not super useful, but I'm sure someone will do something with it. Given that, all the same arguments for sync allocation as IP allocator.
> 
So no one is arguing not to allow that?  Because i would rather make public port and public ip not end user things, thus parts of status, thus reconstructable from external state.
> —
> Reply to this email directly or view it on GitHub.
>